feat: add new conversation filters - WPB-18707

emil-wire opened the pull request:
<!--do not remove this marker, its needed to replace info when ticket title is updated -->
<!--jira-description-action-hidden-marker-start-->

<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-18707" title="WPB-18707" target="_blank"><img alt="Task" src="https://wearezeta.atlassian.net/rest/api/2/universal_avatar/view/type/issuetype/avatar/10818?size=medium" />WPB-18707</a>  [iOS] add filter for unread messages, mentions and replies
  </td></table>
  <br />
 

<!--jira-description-action-hidden-marker-end-->
<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove this marker, its needed to replace info when ticket title is updated -->

<!--do not remove the jira markers to link tickets automatically -->


### Issue

This is my attempt at adding some additional filters to the app. I've tried to stick to the general implementation of the other filter items that already exist. The filters cover unread, mentions and replies. They're hidden behind a feature flag because the design isn't finalized yet. Since this is my first contribution, I'd appreciate all the feedback I can get, especially what tests I should write.

### Testing

Open the app, tap the conversation filters and give it a spin

---

### Checklist

- [x] Title contains a reference JIRA issue number like `[WPB-XXX]`.
- [x] Description is filled and free of optional paragraphs.
- [ ] Adds/updates automated tests.

---

### UI accessibility checklist

_If your PR includes UI changes, please utilize this checklist:_
- [x] Make sure you use the API for UI elements that support large fonts.
- [x] All colors are taken from WireDesign.ColorTheme or constructed using WireDesign.BaseColorPalette.
- [x] New UI elements have Accessibility strings for VoiceOver.

emil-wire committed:
feat: add unread filter WPB-18707

emil-wire committed:
feat: add mentions and replies filter pt.1 WPB-18707

emil-wire committed:
feat: add mentions and replies filter pt2 + predicates

emil-wire committed:
chore: remove project file

emil-wire committed:
chore: lint/format

emil-wire committed:
fix: restore accidentally removed project.pbxproj file

emil-wire committed:
fix: tests (hopefully)

emil-wire committed:
fix: lint

netbe reviewed (commented):
Thanks for the contribution @emil-wire, it will be definitely useful!

2 things:
- I have to kill the app for the flag to take effect
- I am surprised the filter is not applied at the SyncEngine level. We usually add the predicates in ZMConversationListDirectory. 

+ would be cool to add the draft as well

emil-wire committed:
fix: unread icon, feature flag should apply correctly now

emil-wire committed:
fix: moved predicates to the right place, added drafts as well

emil-wire commented on wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/ListContent/ConversationListContentController/ViewModel/ConversationListViewModel.swift:
```diff
@@ -327,16 +353,35 @@ final class ConversationListViewModel: NSObject {
+                filteredSection.items = section.items.filter(filterPredicate)
+                return filteredSection
+            }
+        }
```
great feedback, thanks @netbe! I added them here because
- I didn't know better :D 
- doing it in the UI was quick and dirty :)

emil-wire commented:
@netbe I added drafts, moved the predicates to the right place and hopefully fixed the flag. 
Regarding the tests, could you give me some guidance? Really not sure what to test, how to test. I'm using this as an exercise to learn more about our app and try to understand how it works - so please teach me master :)

emil-wire committed:
added missing strings

emil-wire committed:
fix: random crash that can occur

samwyndham commented on wire-ios/Wire-iOS/Sources/UserInterface/ConversationList/Container/ConversationListViewController.swift:
```diff
@@ -283,6 +291,16 @@ final class ConversationListViewController: UIViewController {
         viewModel.setupObservers()
+
+        // Observe developer flag changes for unread filters
+        NotificationCenter.default.addObserver(
```
issue: I believe when using closure based `NotificationCenter` APIs you need unregister the observation on deinit. To be honest apples docs are not super clear here... Here is an old blog post instead: https://oleb.net/blog/2018/01/notificationcenter-removeobserver/

samwyndham reviewed (dismissed):
Great work! Thank you for contributing! Before merging I think we should either:

- remove the observation of the developer flag
- find a way to make it more granular (I expect KVO) on the UserDefaults object

emil-wire committed:
fix: use ObserverTokens instead closure; some bugs about drafts handling

emil-wire committed:
fix: use ObserverTokens instead closure pt2

netbe commented on wire-ios/Wire-iOS Tests/ConversationList/ConversationListViewModel/ConversationListViewModelFilterIntegrationTests.swift:
```diff
@@ -0,0 +1,470 @@
+
+@testable import Wire
+
+final class ConversationListViewModelFilterIntegrationTests: XCTestCase {
```
issue: seems to be miseed from Xcode proj tests target, mockUnarcchivedConversations to be fixed

netbe reviewed (approved):
looks good, just fix the remained comments tests fixed.

The UI snapshots need to be recorded too with:
IOS_VERSION='18.0'
IPHONE_MODEL='iPhone 14'

emil-wire committed:
fix: KVO, tests

emil-wire commented:
@netbe @samwyndham I've used KVO instead as suggested and fixed up the tests. Added some snapshots using the correct device/os combo too

emil-wire committed:
Merge branch 'develop' into feat/add-filter-for-unread-messages-WPB-18707

emil-wire committed:
fix: remove broken test

emil-wire committed:
Merge origin/feat/add-filter-for-unread-messages-WPB-18707 -> feat/add-filter-for-unread-messages-WPB-18707 | Base: 43bf1fa8995fcf01401b421c59b6fda19220d7c9

emil-wire committed:
fix: lint

emil-wire committed:
fix: tests -.-

emil-wire committed:
fix: lint

emil-wire committed:
fix: snapshots

samwyndham reviewed (approved):
Great work! I'm excited to use this!

emil-wire committed:
Merge branch 'develop' into feat/add-filter-for-unread-messages-WPB-18707

emil-wire commented on wire-ios/Wire-iOS/Sources/UserInterface/MainController/ZClientViewController.swift:
```diff
@@ -163,14 +164,17 @@ final class ZClientViewController: UIViewController {
     var userObserverToken: NSObjectProtocol?
     var conferenceCallingUnavailableObserverToken: Any?
-    var userDidViewSelfProfileToken: NSObjectProtocol?
+    var userDidViewSelfProfileToken: SelfUnregisteringNotificationCenterToken?
```
we do :) I was quite happy about it

emil-wire committed:
Merge branch 'develop' into feat/add-filter-for-unread-messages-WPB-18707

emil-wire committed:
lint

emil-wire committed:
lint
